File: scripts/kokoro-prepare-input.py
#!/usr/bin/env python3
"""Prepare one Kokoro Swift/Core ML input JSON for arbitrary service text.

This script intentionally stops after text/voice preprocessing. The Swift
worker consumes the generated JSON and runs the Core ML synthesis path itself;
normal service requests must not shell out to ``kokoro-bench``.
"""
from __future__ import annotations

import argparse
import json
import os
import sys
from contextlib import redirect_stdout
from pathlib import Path
from typing import Any

ENUM_SIZES = [32, 64, 128, 256, 320, 384, 512]
# packages/contracts MAX_TTS_CHUNK_TOKENS
MAX_CALLER_CHUNK_TOKENS = 450
REPO_ROOT = Path(__file__).resolve().parents[1]


def _runtime_root(arg_root: Path | None) -> Path:
    """Return the Kokoro runtime root for tokenizer/vocab asset lookup.

    Called by ``main`` and service mode before loading vocab or Python Kokoro
    modules. The default is this checkout when it contains repo-local runtime
    assets. ``KOKORO_COREML_ROOT`` and ``--runtime-root`` remain explicit
    overrides for Botnet or generated-runtime layouts.
    """
    if arg_root is not None:
        return arg_root.resolve()
    env_root = os.environ.get("KOKORO_COREML_ROOT", "").strip()
    if env_root:
        return Path(env_root).resolve()
    if (REPO_ROOT / "_kokoro_vocab.json").exists():
        return REPO_ROOT
    return Path.cwd().resolve()


def _ensure_python_runtime_path(root: Path) -> None:
    """Put the runtime root on ``sys.path`` for direct script execution.

    Running ``python scripts/kokoro-prepare-input.py`` sets ``sys.path[0]`` to
    ``scripts/``. Without this helper, the local ``kokoro`` package is invisible
    unless the caller has installed it into the active Python environment.
    """
    root_text = str(root)
    if root_text not in sys.path:
        sys.path.insert(0, root_text)


def _language_for_voice(voice: str) -> str:
    return "b" if voice.startswith("b") else "a"


def _load_vocab(root: Path) -> dict[str, int]:
    home_checkout = Path.home() / "Documents" / "GitHub" / "kokoro-coreml"
    candidates = [
        root / "_kokoro_vocab.json",
        root / "checkpoints" / "config.json",
        root / "outputs" / "hnsf_validation" / "config.json",
        home_checkout / "_kokoro_vocab.json",
        home_checkout / "checkpoints" / "config.json",
        home_checkout / "outputs" / "hnsf_validation" / "config.json",
    ]
    env_root = os.environ.get("KOKORO_COREML_ROOT", "").strip()
    if env_root:
        er = Path(env_root)
        candidates.extend(
            [
                er / "checkpoints" / "config.json",
                er / "outputs" / "hnsf_validation" / "config.json",
            ]
        )
    for candidate in candidates:
        try:
            if candidate.exists():
                config = json.loads(candidate.read_text(encoding="utf-8"))
                if "vocab" in config:
                    return config["vocab"]
        except OSError:
            continue

    raise RuntimeError(
        "Kokoro vocab config not found locally; set KOKORO_COREML_ROOT or run from the kokoro-coreml checkout"
    )


def _prepare_entries(
    *,
    texts: list[str],
    key: str,
    voice: str,
    speed: float,
    pipeline: Any,
    voice_pack: Any,
    vocab: dict[str, int],
) -> list[dict[str, Any]]:
    from kokoro.pipeline import voice_embedding_for_phoneme_string

    entries = []
    for index, text in enumerate(texts):
        phonemes = None
        with redirect_stdout(sys.stderr):
            for _, ps, _ in pipeline(text, voice, speed):
                phonemes = ps
                break

        if not phonemes:
            raise RuntimeError("Kokoro tokenizer returned no phonemes")

        with redirect_stdout(sys.stderr):
            ref_s = voice_embedding_for_phoneme_string(voice_pack, phonemes)

        input_ids = [vocab[p] for p in phonemes if p in vocab]
        input_ids = [0] + input_ids + [0]
        if len(input_ids) > MAX_CALLER_CHUNK_TOKENS:
            raise RuntimeError(
                f"Kokoro chunk exceeds MAX_TTS_CHUNK_TOKENS ({MAX_CALLER_CHUNK_TOKENS}): "
                f"{len(input_ids)} tokens"
            )
        enum_t = next((size for size in ENUM_SIZES if size >= len(input_ids)), ENUM_SIZES[-1])
        padded_ids = input_ids + [0] * max(0, enum_t - len(input_ids))
        attention_mask = [1] * len(input_ids) + [0] * max(0, enum_t - len(input_ids))

        ref_s_list = ref_s.cpu().numpy().flatten().tolist()
        # Emit only what Swift `PreparedKokoroInput` decodes. Phonemes,
        # num_tokens, and the sha256s used to ship as debug metadata;
        # nothing in the worker or runtime read them, so they were dead
        # payload (and phonemes are an unnecessary phonetic mirror of
        # ``text``).
        entries.append({
            "key": key if len(texts) == 1 else f"{key}-{index:03d}",
            "text": text,
            "voice": voice,
            "speed": speed,
            "input_ids": padded_ids,
            "attention_mask": attention_mask,
            "ref_s": ref_s_list,
            "canonical_duration_s": None,
        })
    return entries


def _run_serve(runtime_root: Path | None) -> int:
    root = _runtime_root(runtime_root)
    _ensure_python_runtime_path(root)

    from kokoro import KPipeline

    vocab = _load_vocab(root)
    pipelines: dict[str, Any] = {}
    voice_packs: dict[str, Any] = {}

    def pipeline_for(voice: str) -> tuple[Any, Any]:
        if voice not in pipelines:
            with redirect_stdout(sys.stderr):
                pipeline = KPipeline(lang_code=_language_for_voice(voice))
                voice_pack = pipeline.load_voice(voice)
            pipelines[voice] = pipeline
            voice_packs[voice] = voice_pack
        return pipelines[voice], voice_packs[voice]

    for line in sys.stdin:
        if not line.strip():
            continue
        try:
            request = json.loads(line)
            request_id = str(request["id"])
            texts = request["texts"]
            voice = str(request["voice"])
            speed = float(request["speed"])
            if not isinstance(texts, list) or not all(isinstance(item, str) for item in texts):
                raise ValueError("texts must be a string array")
            pipeline, voice_pack = pipeline_for(voice)
            entries = _prepare_entries(
                texts=texts,
                key="service",
                voice=voice,
                speed=speed,
                pipeline=pipeline,
                voice_pack=voice_pack,
                vocab=vocab,
            )
            response = {"id": request_id, "inputs": entries}
        except Exception as exc:  # noqa: BLE001 - service mode must isolate bad slides.
            request_id = locals().get("request_id", None)
            response = {"id": request_id, "error": str(exc)}
        sys.stdout.write(json.dumps(response, sort_keys=True) + "\n")
        sys.stdout.flush()
    return 0


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--serve", action="store_true")
    parser.add_argument("--text-file", type=Path)
    parser.add_argument("--text-list-file", type=Path)
    parser.add_argument("--output", type=Path)
    parser.add_argument("--key")
    parser.add_argument("--voice")
    parser.add_argument("--speed", type=float)
    parser.add_argument("--runtime-root", type=Path)
    args = parser.parse_args()
    if args.serve:
        return _run_serve(args.runtime_root)
    if (args.text_file is None) == (args.text_list_file is None):
        parser.error("provide exactly one of --text-file or --text-list-file")
    if args.output is None or args.key is None or args.voice is None or args.speed is None:
        parser.error("--output, --key, --voice, and --speed are required outside --serve")

    root = _runtime_root(args.runtime_root)
    _ensure_python_runtime_path(root)

    from kokoro import KPipeline

    texts = (
        [args.text_file.read_text(encoding="utf-8")]
        if args.text_file is not None
        else json.loads(args.text_list_file.read_text(encoding="utf-8"))
    )
    vocab = _load_vocab(root)

    entries = []
    with redirect_stdout(sys.stderr):
        pipeline = KPipeline(lang_code=_language_for_voice(args.voice))
        voice_pack = pipeline.load_voice(args.voice)

    entries = _prepare_entries(
        texts=texts,
        key=args.key,
        voice=args.voice,
        speed=args.speed,
        pipeline=pipeline,
        voice_pack=voice_pack,
        vocab=vocab,
    )

    args.output.parent.mkdir(parents=True, exist_ok=True)
    output = entries[0] if args.text_file is not None else entries
    args.output.write_text(json.dumps(output, sort_keys=True), encoding="utf-8")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/kokoro-prepare-input.mjs
#!/usr/bin/env node

import { existsSync } from 'node:fs';
import { readFile, writeFile } from 'node:fs/promises';
import path from 'node:path';
import process from 'node:process';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { createInterface } from 'node:readline/promises';

const EnumSizes = [32, 64, 128, 256, 320, 384, 512];
// packages/contracts MAX_TTS_CHUNK_TOKENS
const MaxCallerChunkTokens = 450;
const VoiceEmbeddingDim = 256;
const RepoRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const DefaultRuntimeRoot = path.join(RepoRoot, 'packages/kokoro-coreml-runtime');

function parseArgs(argv) {
  const args = new Map();
  const flags = new Set();
  for (let index = 0; index < argv.length; index += 1) {
    const item = argv[index];
    if (!item.startsWith('--')) {
      continue;
    }
    const key = item.slice(2);
    if (key === 'serve') {
      flags.add(key);
      continue;
    }
    args.set(key, argv[index + 1]);
    index += 1;
  }
  return { args, flags };
}

function runtimeRoot() {
  const explicit = process.env.KOKORO_COREML_ROOT?.trim();
  if (explicit) {
    return explicit;
  }
  return DefaultRuntimeRoot;
}

async function loadPhonemizer(root) {
  const candidates = [
    path.join(root, 'kokoro.js/src/phonemize.js'),
    path.join(DefaultRuntimeRoot, 'kokoro.js/src/phonemize.js')
  ];
  for (const candidate of candidates) {
    if (!existsSync(candidate)) {
      continue;
    }
    const module = await import(pathToFileURL(candidate).href);
    if (typeof module.phonemize === 'function') {
      return module.phonemize;
    }
  }
  throw new Error(`Kokoro phonemizer not found under ${root}`);
}

async function loadVocab(root) {
  const candidates = [
    path.join(root, 'outputs/hnsf_validation/config.json'),
    path.join(root, 'checkpoints/config.json'),
    path.join(DefaultRuntimeRoot, 'outputs/hnsf_validation/config.json')
  ];
  for (const candidate of candidates) {
    if (!existsSync(candidate)) {
      continue;
    }
    const parsed = JSON.parse(await readFile(candidate, 'utf8'));
    if (parsed.vocab && typeof parsed.vocab === 'object') {
      return parsed.vocab;
    }
  }
  throw new Error(`Kokoro vocab config not found under ${root}`);
}

async function loadVoiceRows(root, voice) {
  const candidates = [
    path.join(root, 'kokoro.js/voices', `${voice}.bin`),
    path.join(DefaultRuntimeRoot, 'kokoro.js/voices', `${voice}.bin`)
  ];
  for (const candidate of candidates) {
    if (!existsSync(candidate)) {
      continue;
    }
    const bytes = await readFile(candidate);
    if (bytes.byteLength % (VoiceEmbeddingDim * 4) !== 0) {
      throw new Error(`Voice embedding file has invalid byte length: ${candidate}`);
    }
    return new Float32Array(bytes.buffer, bytes.byteOffset, bytes.byteLength / 4);
  }
  throw new Error(`Kokoro voice embedding not found for ${voice}`);
}

function voiceEmbeddingForPhonemes(voiceRows, phonemes) {
  const rowCount = voiceRows.length / VoiceEmbeddingDim;
  const rowIndex = Math.max(0, Math.min(rowCount - 1, phonemes.length - 1));
  const start = rowIndex * VoiceEmbeddingDim;
  return Array.from(voiceRows.slice(start, start + VoiceEmbeddingDim));
}

function languageForVoice(voice) {
  return voice.startsWith('b') ? 'b' : 'a';
}

async function prepareEntries({ texts, key, voice, speed, vocab, voiceRows, phonemize }) {
  const entries = [];
  for (const [index, text] of texts.entries()) {
    const phonemes = await phonemize(text, languageForVoice(voice));
    if (!phonemes) {
      throw new Error('Kokoro phonemizer returned no phonemes');
    }
    const ids = [];
    for (const phoneme of phonemes) {
      const token = vocab[phoneme];
      if (Number.isInteger(token)) {
        ids.push(token);
      }
    }
    const inputIds = [0, ...ids, 0];
    if (inputIds.length > MaxCallerChunkTokens) {
      throw new Error(
        `Kokoro chunk exceeds MAX_TTS_CHUNK_TOKENS (${MaxCallerChunkTokens}): ${inputIds.length} tokens`
      );
    }
    const enumLength = EnumSizes.find((size) => size >= inputIds.length) ?? EnumSizes.at(-1);
    const padded = [...inputIds, ...Array(Math.max(0, enumLength - inputIds.length)).fill(0)];
    const attentionMask = [
      ...Array(inputIds.length).fill(1),
      ...Array(Math.max(0, enumLength - inputIds.length)).fill(0)
    ];
    const refS = voiceEmbeddingForPhonemes(voiceRows, phonemes);
    // Emit only what Swift `PreparedKokoroInput` decodes. Phonemes,
    // num_tokens, and the sha256s used to ship as debug metadata; nothing
    // in the worker or runtime read them, so they were dead payload (and
    // phonemes are an unnecessary phonetic mirror of `text`).
    entries.push({
      key: texts.length === 1 ? key : `${key}-${String(index).padStart(3, '0')}`,
      text,
      voice,
      speed,
      input_ids: padded,
      attention_mask: attentionMask,
      ref_s: refS,
      canonical_duration_s: null
    });
  }
  return entries;
}

async function runServe() {
  const root = runtimeRoot();
  const phonemize = await loadPhonemizer(root);
  const vocab = await loadVocab(root);
  const voices = new Map();
  const lineReader = createInterface({ input: process.stdin, crlfDelay: Infinity });
  for await (const line of lineReader) {
    if (!line.trim()) {
      continue;
    }
    let requestId = null;
    try {
      const request = JSON.parse(line);
      requestId = String(request.id);
      const texts = request.texts;
      const voice = String(request.voice);
      const speed = Number(request.speed);
      if (!Array.isArray(texts) || !texts.every((text) => typeof text === 'string')) {
        throw new Error('texts must be a string array');
      }
      if (!voices.has(voice)) {
        voices.set(voice, await loadVoiceRows(root, voice));
      }
      const inputs = await prepareEntries({
        texts,
        key: 'service',
        voice,
        speed,
        vocab,
        voiceRows: voices.get(voice),
        phonemize
      });
      process.stdout.write(`${JSON.stringify({ id: requestId, inputs })}\n`);
    } catch (error) {
      process.stdout.write(`${JSON.stringify({ id: requestId, error: error instanceof Error ? error.message : String(error) })}\n`);
    }
  }
}

async function runOnce(args) {
  const root = runtimeRoot();
  const textFile = args.get('text-file');
  const textListFile = args.get('text-list-file');
  const output = args.get('output');
  const key = args.get('key');
  const voice = args.get('voice');
  const speed = Number(args.get('speed'));
  if ((!textFile && !textListFile) || (textFile && textListFile) || !output || !key || !voice || !Number.isFinite(speed)) {
    throw new Error('provide exactly one input file plus --output, --key, --voice, and --speed');
  }
  const texts = textFile
    ? [await readFile(textFile, 'utf8')]
    : JSON.parse(await readFile(textListFile, 'utf8'));
  const phonemize = await loadPhonemizer(root);
  const vocab = await loadVocab(root);
  const voiceRows = await loadVoiceRows(root, voice);
  const entries = await prepareEntries({ texts, key, voice, speed, vocab, voiceRows, phonemize });
  await writeFile(output, `${JSON.stringify(textFile ? entries[0] : entries, null, 2)}\n`, 'utf8');
}

const { args, flags } = parseArgs(process.argv.slice(2));
try {
  if (flags.has('serve')) {
    await runServe();
  } else {
    await runOnce(args);
  }
} catch (error) {
  process.stderr.write(`${error instanceof Error ? error.stack ?? error.message : String(error)}\n`);
  process.exitCode = 1;
}
